### PR TITLE
Add static option for Makefiles

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by nalind** [comment]
What's the use case for this?

**Comment by TomSweeneyRedHat** [comment]
We've been getting pinged by a few folks for a static binary.

File: Makefile [hunk 1]
```diff
@@ -15,7 +16,11 @@ LIBSECCOMP_COMMIT := release-2.3
 
 LDFLAGS := -ldflags '-X main.gitCommit=${GIT_COMMIT} -X main.buildInfo=${BUILD_INFO} -X main.cniVersion=${CNI_COMMIT}'
 
-all: buildah imgtype docs
+all: buildah imgtype docs static
+
+.PHONY: static
+static: *.go imagebuildah/*.go bind/*.go chroot/*.go cmd/buildah/*.go docker/*.go pkg/cli/*.go pkg/parse/*.go unshare/*.c unshare/*.go util/*.go
+	$(GO) build -ldflags '-X main.gitCommit=ba976ed3 -X main.buildInfo=1533330547 -X main.cniVersion=v0.6.0' -ldflags "-extldflags '-static'" -o buildah.static -tags "$(TAGS) $(SELINUXTAG) containers_image_ostree_stub containers_image_openpgp exclude_graphdriver_devicemapper" ./cmd/buildah
```
Comment (nalind) [inline Makefile]: Can we find a cleaner way to do this?  Directly specifying the values that we're also storing in variables is bound to create mismatches when we eventually forget to update both copies when making a change to one or the other.

Comment (rhatdan) [reply Makefile]: Suggestion on how?

Comment (nalind) [reply Makefile]: Well, comparing the output of `make -n buildah` and `make -n static`, I see three build tags being added and a second `-ldflags` argument being used.

Adding a variable, similar to $(TAGS) but with no default value, that gets incorporated into $(LDFLAGS), and making the value of the `-o` argument to the `go build` command a variable with a default value of `buildah` would make it possible for the `static` build target to do its work by recursively invoking `$(MAKE)` with `TAGS="$(TAGS) addedtags`, and specifying values for the new variables that we use for holding the additional LDFLAGS and the binary name.

**Comment by rhatdan** [comment]
@nalind I rewrote the Makefile to handle SECURITYTAGS and STORAGETAGS and I added modifications along the lines of your suggestions, for handling of BUILDAH.STATIC

**Comment by rhatdan** [comment]
Finally builds successfully
@nalind @TomSweeneyRedHat @giuseppe @pixdrift @boaz1337 @vrothberg @umohnani8 PTAL

File: Makefile [hunk 2]
```diff
@@ -1,27 +1,37 @@
-AUTOTAGS := $(shell ./btrfs_tag.sh) $(shell ./btrfs_installed_tag.sh) $(shell ./libdm_tag.sh) $(shell ./ostree_tag.sh) $(shell ./selinux_tag.sh)
-TAGS ?= seccomp
+SELINUXTAG := $(shell ./selinux_tag.sh)
+STORAGETAGS := $(shell ./btrfs_tag.sh) $(shell ./btrfs_installed_tag.sh) $(shell ./libdm_tag.sh) $(shell ./ostree_tag.sh)
+SECURITYTAGS ?= seccomp $(SELINUXTAG)
```
Comment (vrothberg) [inline Makefile]: I am afraid that removing `TAGS` could break some builds in the wild that use `TAGS`. To remain backwards compat, we could add:
```
TAGS ?= $(SECURITYTAGS) $(STORAGETAGS)
[...]
BUILDFLAGS := -tags "$(TAGS)"
```

Would this work or am I missing something?

**Comment by vrothberg** [comment]
:+1: for offering the option to have static builds. Just a minor nitpick and the request for a commit message. The changes add some new variables that are worth mentioning.

File: Makefile (hunk 2 above)
Comment (rhatdan) [reply Makefile]: Added.

File: install.md [hunk 3]
```diff
@@ -182,7 +182,7 @@ Then to install Buildah on Ubuntu follow the steps in this example:
   export GOPATH=`pwd`
   git clone https://github.com/containers/buildah ./src/github.com/containers/buildah
   cd ./src/github.com/containers/buildah
-  PATH=/usr/lib/go-1.10/bin:$PATH make runc all TAGS="apparmor seccomp"
+  PATH=/usr/lib/go-1.8/bin:$PATH make runc all SECURITYTAGS="apparmor seccomp"
```
Comment (TomSweeneyRedHat) [inline install.md]: should this still be 1.10 and not back to 1.8?

Comment (rhatdan) [reply install.md]: Yes

File: .papr.sh [hunk 4]
```diff
@@ -36,6 +38,6 @@ go get github.com/onsi/gomega/...
 # short-commit-subject validation test, so tell git-validate.sh to only check
 # up to, but not including, the merge commit.
 export GITVALIDATE_TIP=$(cd $GOSRC; git log -2 --pretty='%H' | tail -n 1)
-make -C $GOSRC install.tools runc all validate test-unit test-integration TAGS="seccomp"
-env BUILDAH_ISOLATION=chroot make -C $GOSRC test-integration TAGS="seccomp"
-env BUILDAH_ISOLATION=rootless make -C $GOSRC test-integration TAGS="seccomp"
+make -C $GOSRC install.tools runc all validate test-unit test-integration static
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply .papr.sh]: SECURITYTAGS defaults to seccomp, so no need for this.

File: chroot/run.go [hunk 5]
```diff
@@ -955,6 +955,20 @@ func setRlimits(spec *specs.Spec, onlyLower, onlyRaise bool) error {
...
+		return errors.Wrapf(err, "error checking if directory %q was bound read-only", mntpoint)
+	}
+	if fs.Flags&unix.ST_RDONLY == 0 {
+		if err := unix.Mount(mntpoint, mntpoint, "bind", flags|unix.MS_REMOUNT, ""); err != nil {
+			return errors.Wrapf(err, "error remounting %s in mount namespace read-only", mntpoint)
+		}
+	}
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply chroot/run.go]: Not there any longer.

**Comment by rhatdan** [comment]
@nalind @TomSweeneyRedHat @giuseppe @pixdrift @boaz1337 @vrothberg @umohnani8 PTAL

**Comment by umohnani8** [comment]
LGTM

**Comment by rhatdan** [comment]
Sadly this does not work now.
```
 make static
make SECURITYTAGS="seccomp selinux" STORAGETAGS="containers_image_ostree_stub containers_image_openpgp exclude_graphdriver_devicemapper " EXTRALDFLAGS='-ldflags "-extldflags '-static'"' BUILDAH=buildah.static binary
make[1]: Entering directory '/home/dwalsh/go/src/github.com/containers/buildah'
go build -ldflags '-X main.gitCommit=604214fd -X main.buildInfo=1542728105 -X main.cniVersion=v0.7.0-alpha1' -ldflags "-extldflags -static" -o buildah.static -tags "seccomp selinux containers_image_ostree_stub containers_image_openpgp exclude_graphdriver_devicemapper " ./cmd/buildah
# github.com/containers/buildah/cmd/buildah
/bin/ld: /tmp/go-link-317787411/000002.o: in function `mygetgrouplist':
/builddir/build/BUILD/go/src/os/user/getgrouplist_unix.go:16: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/bin/ld: /tmp/go-link-317787411/000001.o: in function `mygetgrgid_r':
/builddir/build/BUILD/go/src/os/user/cgo_lookup_unix.go:38: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/bin/ld: /tmp/go-link-317787411/000001.o: in function `mygetgrnam_r':
/builddir/build/BUILD/go/src/os/user/cgo_lookup_unix.go:43: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/bin/ld: /tmp/go-link-317787411/000001.o: in function `mygetpwnam_r':
/builddir/build/BUILD/go/src/os/user/cgo_lookup_unix.go:33: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/bin/ld: /tmp/go-link-317787411/000001.o: in function `mygetpwuid_r':
/builddir/build/BUILD/go/src/os/user/cgo_lookup_unix.go:28: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/bin/ld: /tmp/go-link-317787411/000014.o: in function `_cgo_18049202ccd9_C2func_getaddrinfo':
/tmp/go-build/cgo-gcc-prolog:49: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
make[1]: Leaving directory '/home/dwalsh/go/src/github.com/containers/buildah'
```

**Comment by rhatdan** [comment]
I will merge for now, since the static version works, I am running it in a container without only buildah.static.

@rh-atomic-bot r+

**Pull request closed without merging**